samcli/commands/local/lib/intrinsic_resolver/intrinsic_property_resolver.py

```diff
@@ -0,0 +1,759 @@
+        self.conditions = conditions or self.template.get("Conditions", {})
+
+        self.symbol_resolver = symbol_resolver
+        self.symbol_resolver.resources = self.resources
+        self.symbol_resolver.parameters = self.parameters
+
+        self.intrinsic_key_function_map = {
+            IntrinsicResolver.FN_JOIN: self.handle_fn_join,
+            IntrinsicResolver.FN_SPLIT: self.handle_fn_split,
+            IntrinsicResolver.FN_SUB: self.handle_fn_sub,
+            IntrinsicResolver.FN_SELECT: self.handle_fn_select,
+            IntrinsicResolver.FN_BASE64: self.handle_fn_base64,
+            IntrinsicResolver.FN_FIND_IN_MAP: self.handle_find_in_map,
+            IntrinsicResolver.FN_TRANSFORM: self.handle_fn_transform,
+            IntrinsicResolver.FN_GET_AZS: self.handle_fn_get_azs,
+            IntrinsicResolver.REF: self.handle_fn_ref,
+            IntrinsicResolver.FN_GET_ATT: self.handle_fn_getatt,
+            IntrinsicResolver.FN_IMPORT_VALUE: self.handle_fn_import_value
+        }
+
+        self.conditional_key_function_map = {
+            IntrinsicResolver.FN_AND: self.handle_fn_and,
+            IntrinsicResolver.FN_OR: self.handle_fn_or,
+            IntrinsicResolver.FN_IF: self.handle_fn_if,
+            IntrinsicResolver.FN_EQUALS: self.handle_fn_equals,
+            IntrinsicResolver.FN_NOT: self.handle_fn_not
+        }
+
+    def intrinsic_property_resolver(self, intrinsic, parent_function="template"):
+        """
+        This resolves the intrinsic of the format
+        {
+            intrinsic: dict
+        } by calling the function with the relevant intrinsic function resolver.
+
+        This also supports returning a string, list, boolean, int since they may be intermediate steps in the recursion
+        process. No transformations are done on these.
+
+        By default this will just return the item if non of the types match. This is because of the function
+        resolve_all_attributes which will recreate the resources by processing every aspect of resource.
+
+        Parameters
+        ----------
+        intrinsic: dict, str, list, bool, int
+            This is an intrinsic property or an intermediate step
+        parent_function: str
+            In case there is a missing property, this is used to figure out where the property resolved is missing.
+
+        Return
+        ---------
+        The simplified version of the intrinsic function. This could be a list,str,dict depending on the format required
+        """
+        if intrinsic is None:
+            raise InvalidIntrinsicException("Missing Intrinsic property in {}".format(parent_function))
+
+        if any(isinstance(intrinsic, object_type) for object_type in [string_types, list, bool, int]):
+            return intrinsic
+
+        if intrinsic == {}:
+            return intrinsic
+
+        if not isinstance(intrinsic, dict):
+            raise InvalidIntrinsicException(
```

Comment:
This would never happen right? You are explicitly checking for other values above. Won't this be a dict?

The message also says "It is not a int, list, str, or dict" but you are checking for if it is not a dict.

Reply:
The only case I was consider was if an object was there. I'll remove it since we are assuming that it's valid CloudFormation read from a file.